codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt

```diff
@@ -1,56 +1,304 @@
+                serializeUnion(Context(writerExpression, context.valueExpression, target))
+            }
+            else -> TODO(target.toString())
+        }
+    }
+
+    private fun determineTimestampFormat(shape: MemberShape): TimestampFormatTrait.Format =
+        shape.getMemberTrait(model, TimestampFormatTrait::class.java).orNull()?.format
+            ?: TimestampFormatTrait.Format.DATE_TIME
+
+    private fun RustWriter.structWriter(context: MemberContext, inner: RustWriter.(String) -> Unit) {
+        val prefix = context.shape.getTrait<XmlNameTrait>()?.value ?: context.shape.memberName
+        safeName("scope").also { scopeName ->
+            Attribute.AllowUnusedMut.render(this)
+            rust("let mut $scopeName = ${context.writerExpression}.prefix(${prefix.dq()});")
+            inner(scopeName)
+        }
+    }
+
+    private fun RustWriter.serializeCollection(memberContext: MemberContext, context: Context<CollectionShape>) {
+        val flat = memberContext.shape.getTrait<XmlFlattenedTrait>() != null
+        val memberOverride = when (val override = context.shape.member.getTrait<XmlNameTrait>()?.value) {
+            null -> "None"
+            else -> "Some(${override.dq()})"
+        }
+        val itemName = safeName("item")
+        safeName("list").also { listName ->
+            rust("let mut $listName = ${context.writerExpression}.start_list($flat, $memberOverride);")
+            rustBlock("for $itemName in ${context.valueExpression.asRef()}") {
+                val entryName = safeName("entry")
+                rust("let mut $entryName = $listName.entry();")
+                val targetShape = model.expectShape(context.shape.member.target)
+                serializeMemberValue(
+                    MemberContext(entryName, ValueExpression.Reference(itemName), context.shape.member),
+                    targetShape
+                )
+            }
+            rust("$listName.finish();")
+        }
+    }
+
+    private fun RustWriter.serializeMap(memberContext: MemberContext, context: Context<MapShape>) {
+        val flat = memberContext.shape.getTrait<XmlFlattenedTrait>() != null
+        val entryKeyName = (context.shape.key.getTrait<XmlNameTrait>()?.value ?: "key").dq()
+        val entryValueName = (context.shape.value.getTrait<XmlNameTrait>()?.value ?: "value").dq()
+        safeName("map").also { mapName ->
+            val keyName = safeName("key")
+            val valueName = safeName("value")
+            rust("let mut $mapName = ${context.writerExpression}.start_map($flat, $entryKeyName, $entryValueName);")
+            rustBlock("for ($keyName, $valueName) in ${context.valueExpression.asRef()}") {
+                val keyTarget = model.expectShape(context.shape.key.target)
+                val keyExpression = when (keyTarget.hasTrait<EnumTrait>()) {
+                    true -> "$keyName.as_str()"
+                    else -> keyName
+                }
```

Comment:
at some point I had a helper that literally just did this for enums, but I guess I deleted it